### ISSUE TITLE
About VLMs

### ISSUE THREAD

**Issue by YUECHE77** [CSIPlab/SLUG#5]
Hi!

We are really interested in your work! While trying to reproduce the VLM results, I noticed that LLaVA-v1.5 uses `openai/clip-vit-large-patch14-336` as its vision encoder, which is different from the LAION version.

Does this mean that, in order to reproduce LLaVA's results in Table 3, we also need to fine-tune `openai/clip-vit-large-patch14-336`? If so, could you please provide the fine-tuning script so that we can perfectly reproduce your results?

Meanwhile, could you please also upload the tar files for the other identities? For example, in Table 3: KANYE WEST, TOM CRUISE,  BARACK OBAMA, LADY GAGA, and BRUCE LEE. Additionally, in Table 6 (Appendix): BASKETBALL, BEACH, CASTLE, REVOLVER, RIFLE, SCHOOLBUS, and SUNGLASSES. That would be very helpful!

Thank you so much!!

**Comment by YUECHE77** [comment]
In the meantime, regarding the **Forget Accuracy** evaluation in Table 3, the paper mentions:

"We sample 10 different targeted identities from the CelebA, with 100 images per identity, resulting in 1000 images in total, to create a comprehensive validation set."

I understand the process of sampling 10 different targeted identities from CelebA, which is straightforward. However, I assume you are not using the images from CelebA itself to construct the **comprehensive validation set**, since CelebA only contains about `10–20` images per person.

Could you please clarify which dataset you are sampling from, or could you upload the 1000 images used for this validation set?

I would really appreciate it if you could confirm this point. Thanks again!

**Comment by YtengTan** [comment]
Hi @YUECHE77,

Thanks for your interest in our work. Regarding your questions I have provided instructions/clarifications below.
Feel free to get back to me, if there's anything remain unclear to you.

### VLM experiments
Our scrips for CLIP experiments are based on [open_clip](https://github.com/mlfoundations/open_clip) implementation. LLaVA-1.5 uses [huggingface diffusers](https://huggingface.co/docs/diffusers/en/index) implementation, specifically, ["openai/clip-vit-large-patch14-336"](https://huggingface.co/openai/clip-vit-large-patch14-336) as its vision encoder.  
1. To compute forget & retain gradients for huggingface models, run `bash scripts/run_compute_grad_hf.sh` (uncomment [line-8](https://github.com/CSIPlab/SLUG/blob/2c4ef98a652a2c1fe2acdca913861b4b04a85ad7/scripts/run_compute_grad_hf.sh#L8) for LLaVA-1.5 CLIP).
2. To search pareto-front layers and step-sizes use [SLUG/notebooks/experiment_vision_language_search.ipynb](https://github.com/CSIPlab/SLUG/blob/25f8d3cca06d3960b02d594d19b6c4ad16beb7cb/notebooks/experiment_vision_language_search.ipynb).
3. Our VLM evaluation is based on [lmms-eval v0.3.0](https://github.com/EvolvingLMMs-Lab/lmms-eval/tree/v0.3.0) code base. To evaluate unlearned VLMs, download and install `lmms-eval` following the instructions. Copy and paste [SLUG/src/lmms-eval/lmms_eval/tasks](https://github.com/CSIPlab/SLUG/tree/25f8d3cca06d3960b02d594d19b6c4ad16beb7cb/src/lmms-eval/lmms_eval/tasks) folders to your downloaded [lmms-eval/lmms_eval/tasks](https://github.com/EvolvingLMMs-Lab/lmms-eval/tree/9a895c4424b71c65745cf260c6a6b25b25c3e428/lmms_eval/tasks). Then, run 
```text
python3 -m accelerate.commands.launch \
    --num_processes=1 \
    -m lmms_eval \
    --model llava_hf \
    --model_args "ytan-ucr/llava-1.5-unlearned-elon-musk,device_map=cuda:0" \
    --tasks mu_elon \
    --batch_size 1 \
    --log_samples \
    --output_path ./logs/
```
you can pass `--model_args "pretrained=llava-hf/llava-1.5-7b-hf,device_map=cuda:0"` to evaluate original [llava-1.5](https://huggingface.co/llava-hf/llava-1.5-7b-hf) accuracy on identifying Elon Musk images. For evaluation on other identity, pass `--tasks mu_taylor,mu_mark` etc., and pass `--tasks mme,gqa,mmbench` for VLM utility evaluation.

> Note that due to the storage limitation, we have only provided the complete model weight for `llava-1.5-unlearned-elon-musk` on our [huggingface](https://huggingface.co/ytan-ucr/models). For other identities (e.g., [Jeff Bezos](https://huggingface.co/ytan-ucr/llava-1.5-unlearned-jeff-bezos/tree/main), [Tom Cruise](https://huggingface.co/ytan-ucr/llava-1.5-unlearned-tom-cruise/tree/main)), we have uploaded only the tuned layer weights: `model-mu.safetensors` and a corresponding `model.safetensors.index.json`. If you wish to run evaluation on those identities, you will need to download the complete [llava-1.5](https://huggingface.co/llava-hf/llava-1.5-7b-hf) model directory, and move `model-mu.safetensors` and `model.safetensors.index.json` under each `llava-1.5-unlearned-[celeb id]` to the llava-1.5 model directory, so that evaluation script loads updated layer weights for unlearned identity.


### Other tar files for supplementary experiments
We have uploaded other identity and object tar files [here](https://drive.google.com/drive/folders/1K8DCnw3B56hUcxF-8SYWYo-AY1uLAWC1). If you wish to run on customized data set, you can use [SLUG/src/clip/a0_create_tar.py](https://github.com/CSIPlab/SLUG/blob/main/src/clip/a0_create_tar.py) to create your own tar files.


### Clarification on VLM datasets
The 10 identities are sampled from CelebA label space based on their appearing frequency, this intuition is to avoid less frequent appeared celebrity that LLaVA/CLIP has poor classification performance on. We customize our own 1000 images data set on the selected 10-identity, on a basis that original LLaVA/CLIP have high accuracy on these data. The dataset is publicly available at [https://huggingface.co/ytan-ucr/datasets](https://huggingface.co/ytan-ucr/datasets).

**Comment by YUECHE77** [comment]
Hi @YtengTan ,

Thank you so much for your prompt reply and the detailed instructions!!! I really appreciate it.

However, I still have two follow up questions regarding the VLM experiments:

1. When training LLaVA-1.5's vision encoder ([openai/clip-vit-large-patch14-336](https://huggingface.co/openai/clip-vit-large-patch14-336)), why is the celebrity's name used as [text input](https://github.com/CSIPlab/SLUG/blob/f298ac3b5a287e4200b69a8ca695d99f796b93e2/src/clip/get_gradients_hf.py#L197) instead of the paired image description from the [dataset](https://github.com/CSIPlab/SLUG/blob/f298ac3b5a287e4200b69a8ca695d99f796b93e2/src/clip/unlearn/calc_grad.py#L43)? In other words, why are the training procedures different between the LAION version CLIP and LLaVA-1.5's vision encoder? Is there a specific reason for this inconsistency?

2. For the forget set when training the LAION version CLIP ([script](https://github.com/CSIPlab/SLUG/blob/f298ac3b5a287e4200b69a8ca695d99f796b93e2/src/clip/unlearn/calc_grad.py#L68)), the loss is: 
```python
(-1) * CosineEmbeddingLoss()(image_features, text_features, torch.ones(images.size(0)).to(device))
```
This makes sense, since we want to minimize the cosine similarity between image and text in the forget set. However, when training LLaVA-1.5's vision encoder ([script](https://github.com/CSIPlab/SLUG/blob/f298ac3b5a287e4200b69a8ca695d99f796b93e2/src/clip/get_gradients_hf.py#L217)), we don't have the negative sign. Could you clarify why the loss functions are different? Is this intentional, or did I miss something?

Looking forward to your clarification, and thanks again for your previous response!

**Comment by YUECHE77** [comment]
Hi @YtengTan ,

I followed the instruction but failed to reproduce your VLM results (the forget accuracy).

Could you please do a quick double check and run from your end to make sure all the VLM scripts are correct?

Thanks!
